ports/nrf/modules/machine/adc.c

```diff
@@ -112,14 +152,102 @@ STATIC mp_obj_t machine_adc_make_new(const mp_obj_type_t *type, size_t n_args, s
 /// Read adc level.
 mp_obj_t machine_adc_value(mp_obj_t self_in) {
     machine_adc_obj_t *self = self_in;
-    return MP_OBJ_NEW_SMALL_INT(hal_adc_channel_value(&self->adc->config));
+
+    int16_t value;
+
+#if NRF51
+    nrfx_adc_channel_t channel_config = {
+        .config.config.resolution         = NRF_ADC_CONFIG_RES_8BIT,
+        .config.config.input              = NRF_ADC_CONFIG_SCALING_INPUT_TWO_THIRDS,
+        .config.config.reference          = NRF_ADC_CONFIG_REF_VBG,
+        .config.config.ain                = self->ain,
+        .config.config.external_reference = ADC_CONFIG_EXTREFSEL_None << ADC_CONFIG_EXTREFSEL_Pos // Currently not defined in nrfx/hal.
+    };
+
+    nrfx_adc_sample_convert(&channel_config, &value);
+#else // NRF52
+    nrfx_saadc_sample_convert(self->id, &value);
+#endif
+
+    return MP_OBJ_NEW_SMALL_INT(value);
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(mp_machine_adc_value_obj, machine_adc_value);
 
+#if NRF51
+
+#define ADC_REF_VOLTAGE_IN_MILLIVOLT (1200) // Reference voltage in mV (1.2V).
+#define ADC_PRE_SCALING_MULTIPLIER   (3)    // VDD 1/3 prescaling as input. Hence, multiplied by 3 to get the value of the battery voltage.
+
+#else // NRF52
+
+#define ADC_REF_VOLTAGE_IN_MILLIVOLT (600)  // Reference voltage in mV (0.6V).
+#define ADC_PRE_SCALING_MULTIPLIER   (6)    // VDD 1/6 prescaling as input. Hence, multiplied by 6 to get the value of the battery voltage.
+
+#endif
+
+#define DIODE_VOLT_DROP_MILLIVOLT    (270)  // Voltage drop over diode.
+
+#define BATTERY_MILLIVOLT(VALUE) \
+        ((((VALUE) * ADC_REF_VOLTAGE_IN_MILLIVOLT) / 255) * ADC_PRE_SCALING_MULTIPLIER)
+
+static uint8_t battery_level_in_percent(const uint16_t mvolts)
+{
+    uint8_t battery_level;
+
+    if (mvolts >= 3000) {
```

Comment:
I see this algorithm was already in the nrf port, but I think it would be nice to explain it a bit. What are those magic numbers and where do they come from? For what battery technology is this?

Reply:
Yeah, this is a blury API. Its for coincell, and depends on a diode drop in the calculation. I personally find this bonus API usefull when using nordic kits that can run coincell, but its not generic enough. I believe this can be improved in the future, or removed and let python handle the ADC measurement itself. If so, then there should also be an API parameter to select internal reference.

Reply:
Maybe put the battery technology in the name? Something like `coincell_level_in_percent` would make it clear this is only for coin cells.